### ISSUE TITLE
Simplify Bootstrappers

### ISSUE THREAD

**Issue by gep13** [cake-build/resources#53]
The current default bootstrappers (i.e. the ones in this repository, prior to personal modifications) do a number of things, these include:

- Makes sure that the tools folder exists
- Makes sure that the packages.config exists (if it doesn't, the default packages.config file is downloaded from this repository)
- Checks to see if NuGet.exe is present, if not, download it
- Check to see if packages.config file has changed since last build execution, if it has, delete contents of tools folder
- Restore tools from packages.config file
- Update packages.config file hash
- If there is a addins/packages.config file, restore addins
- If there is a modules/packages.config file, restore modules
- Test to see if cake.exe is present in tools folder, if not, throw error
- Prepare Cake Arguments, and invoke Cake.exe

**Comment by gep13** [comment]
The plan for the 0.26.0 will be to simplify the bootstrappers, so that they aren't doing as much work, and instead using other features that are now present in Core Cake.

**Comment by gep13** [comment]
As an example, the MD5 hashing that was added to the bootstrappers was only required due to the fact that out of the box, Cake.exe didn't handle changing the version number of an addin/tool.  Now that this works out of the box using the In-Process NuGet Client, this is not required in the bootstrapper.

**Comment by gep13** [comment]
Also, now that all external dependencies (addins/modules/tools) can be installed via pre-processor directives, I think the recommendation has to be to NOT use packages.config files.

**Comment by gep13** [comment]
@cake-build/cake-team thoughts?

**Comment by gep13** [comment]
@jnm2 what are your thoughts here?

**Comment by devlead** [comment]
@gep13 yes, I'll take a look at this and propose a PR

**Comment by gep13** [comment]
@devlead I am leaning towards putting the pinned Cake version number into the cake.config file. Thoughts?

**Comment by patriksvensson** [comment]
@gep13 Does that mean we need to start parsing the cake.config file in the bootstrapper?

**Comment by mholo65** [comment]
@patrik @gep13 *shrug* parsing ini-file in  bootstrapper seems like an awful amount of work. Can we do better?

And @gep13, not downloading `Cake` via NuGet, meaning ”less” downloads as we read stats today.

**Comment by gep13** [comment]
We wouldn't need to parse the entire contents. A simple regex would determine the version number. It would mean that the bootstrapper could become an artifact that doesn't need to change, and can be something that we can provide for each release of Cake going forward.

@mholo65 surely using the direct nuget download URL would still count towards overall download count.

**Comment by patriksvensson** [comment]
@mholo65 Yes, but if bootstrappers download from GitHub we can get stats from the GitHub API.

Personally, I would want to see at `#cake 0.26.0` attribute that specifies the lowest Cake version that the script works with.

**Comment by devlead** [comment]
IMHO parsing ini files is to much responsibility for the bootstrappers.

Cake preprocessor sounds like a good idea, should raise separate issue about that.

**Comment by jnm2** [comment]
🎉 You rock! I had this bookmarked to on my todo list of issues to start here: https://gitter.im/cake-build/cake?at=5a46600468d092bb62018663

My current thinking on the ideal bootstrapper:

 1. Reads the .cake file specified as the first non-switch parameter
 2. Scans for a `#cake version=0.24 dir=tools/Cake source=https://www.nuget.org/api/v2` line, if any exist
 3. HTTP-gets`{source}/package/Cake/{version}`and unzips as the download streams, decompressing to files at `{dir}`, using defaults if necessary
 4. Passes the unaltered argument list to Cake.exe with the exception of translating known parameters from PowerShell style (-ArgName:Value) so that tab completion works.

Yes, that is a bit of work for the bootstrapper. But it enables a world where all the bootstrapper is then considered *uncustomizable*– customizations either go in directives like `#cake` and `#module`, or in cake.config. The bootstrapper should be an **opaque black box** released alongside each release of the Cake binaries which may once in a while need to be updated in your repo when you upgrade your Cake version for compatibility with Cake's arguments and new .cake directives or other capabilities (noted in the release notes when this happens).

tl;dr You should only need to touch the bootstrapper extremely rarely, and when you do, you should be able to replace it entirely with the most recent version without having to think about what's inside the file.

If there is any part of that picture that you like, I'll drop everything and help make it happen. I've done closely similar things in the past and do enjoy PowerShell. I'd have to do some learning for Bash.

**Comment by jnm2** [comment]
Before you say it's too much work for the bootstrapper to read a `#cake` directive (if any) to locate a pinned version (if any), a directory (if any) and a package source (if any), would you please let me try? 😃
Even if it results in a long bootstrapper, that is not entirely a bad thing. It further incentivizes people to look to #directives and cake.config instead of customizing the bootstrapper which [has never worked well](https://gitter.im/cake-build/cake?at=5a46579f68d092bb62015baa).

**Comment by jnm2** [comment]
Before you say it's too much work for the bootstrapper to read `#cake` directives (if any) to locate a pinned version (if any), a target directory (if any) and a package source (if any), would you please let me try? 😃
Even if it results in a long bootstrapper, that is not entirely a bad thing. It further incentivizes people to look to #directives and cake.config instead of customizing the bootstrapper which [has never worked well](https://gitter.im/cake-build/cake?at=5a46579f68d092bb62015baa).

I don't want to miss the chance while you're all talking about it to make the bootstrapper as ideal as it can be!

Also, we might consider having the bootstrapper with the way Chocolatey's works: a tiny file which HTTPs-downloads and executes PowerShell (or potentially a C# binary).

```ps1
Set-ExecutionPolicy Bypass -Scope Process -Force; iex ((New-Object System.Net.WebClient).DownloadString('https://chocolatey.org/install.ps1'))
```

**Comment by gep13** [comment]
>@devlead said...
IMHO parsing ini files is to much responsibility for the bootstrappers.

How is parsing the ini file any different from parsing for a pre-processor directive?  Both would have to happen before the Cake.exe/Cake.CoreClr.dll could be invoked.

**Comment by devlead** [comment]
@gep13 I was thinking about
> #cake 0.26.0 attribute that specifies the lowest Cake version that the script works with.

**Comment by devlead** [comment]
I think by letting the bootstrappers do less and cake do more - the need to modify them decreases dramatically. Also makes it easier to have the scripts more portable. As PowerShell should preferably work on 2.0+ and core. Bash script should have as few dependencies as possible.

If the bootstrappers stays away from argument parsing and things like that and just passed along to Cake, many of issues people have had would be eradicated. Would also remove the differences between platforms which really confuses people today. 

By reducing it to 1 moving part people won't forget to pin version as it's pinned there by default and they won't get new version until they change there, and nothing will work if they don't commit that file.

**Comment by jnm2** [comment]
To me, the ability to pin the cake version via `#cake version=0.24` is worth whatever it takes.

**Comment by gep13** [comment]
> #cake 0.26.0 attribute that specifies the lowest Cake version that the script works with.

So does this mean that Cake itself will need to download a newer version if it needs to? Or how do you determine what version of Cake is required for first execution of script?

**Comment by patriksvensson** [comment]
@gep13 No downloading would be made. Simply refuse to execute the script.

**Comment by jnm2** [comment]
This proof-of-concept is handling everything I can throw at it:

<details>

```ps1
# TODO: parse arguments
$buildScriptPath = 'build.cake'

$directiveOptions = @{}

try {
    foreach ($optionMatch in
        select-string -Path $buildScriptPath '^\s*#cake\s+(?<options>.*)$' -AllMatches |
        foreach Matches |
        foreach { $_.Groups['options'].Value } |
        select-string '(?<key>version|dir|source)\s*=\s*(?<value>[^"\s]+|"([^"]|"")*")' -AllMatches |
        foreach Matches |
        where { $_ -ne $null }
    ) {
        $optionKey = $optionMatch.Groups['key'].Value
        $optionValue = $optionMatch.Groups['value'].Value

        if ($directiveOptions.ContainsKey($optionKey)) {
            throw "`"#cake $optionKey`" must only appear once in '$buildScriptPath'."
        }

        $directiveOptions.Add($optionKey, $optionValue);
    }
}
catch [System.Management.Automation.ItemNotFoundException] {
    throw "Cannot find build script at '$buildScriptPath'."
}

# Print to console
$directiveOptions

$pinnedVersion = $directiveOptions.version
$targetDir = $directiveOptions.dir
$nugetSource = $directiveOptions.source
```

</details>

Tested against ‘interesting’ .cake file:
```cake
#cake version=1 someProperty="testing\1\2\3"

 #CAKE SOURCE = "\\myserver\myfeed"
```

**Comment by gep13** [comment]
@patriksvensson so how would a user get Cake initially then?

**Comment by patriksvensson** [comment]
@gep13 Like today. My proposal is just to verify that a minimum Cake version is running. I want to bootstrappers super small and super lean. More or less download a version of Cake.

**Comment by gep13** [comment]
So today, if there is no Cake downloaded, the bootstrapper will either download the latest version, if there is no packages.config file. If there is a packages.config file it will download the version listed in there. Without first parsing the cake.config files or looking for pre-processor directive, we will lose this latter part, which i think is important.

**Comment by patriksvensson** [comment]
@gep13 I don't see the problem. You just specify the version of Cake you want to download in the bootstrapper script, like [here](https://github.com/spectresystems/spectre.commandline/blob/master/build.ps1#L34).

**Comment by jnm2** [comment]
@patriksvensson Customizing a bootstrapper, even for something as simple as pinning the Cake version, is problematic when you use multiple bootstrappers for multiple platforms. It gets overlooked (at NUnit it cost me a lot of time) and it is a maintenance burden.

**Comment by devlead** [comment]
Or like [here](https://github.com/LitJSON/litjson/blob/develop/build.ps1#L).

**Comment by devlead** [comment]
@jnm2 having complex bootstrappers have been a burden too.

**Comment by jnm2** [comment]
Once you no longer have people trying to customize them, then the complexity is no longer a burden for Cake users. Then we can treat the bootstrapper like a black box.

**Comment by patriksvensson** [comment]
Well, I don't agree. The bootstrapper is supposed to be customized. We should make them as simple as possible and put more functionality in Cake itself if it's required.

For example, the new `--bootstrap` switch could be used to download the version of Cake specified by the `#cake [version]` pre-processor.

**Comment by jnm2** [comment]
So long as `#cake` works and I don't have to customize any OS-specific bootstrapper, I'm happy. I'm curious though: besides the reason that you've done it this way in the past, why do you say the bootstrapper is supposed to be customized?

**Comment by devlead** [comment]
@jnm2 We didn't have everything in place before, now for example #addin & #tool directive handle versioning a lot better. We don't need nuget.exe anymore. We've can bootstrap modules, and as @patriksvensson states the `--boostrap` functionality recently introduced, could be an embryo for future needs. 

Also anything removed from bootstrapper to Cake per definition will make the bootstrappers less OS specific.

**Comment by jnm2** [comment]
@devlead Yes, removing that stuff is great and I'm certainly not advocating leaving it in, haha! I am, however, advocating taking *all* of it out and replacing it with the simplest possible downloader that can understand a #cake version directive and download the correct version the first time.

[This is what it looks like.](https://gist.github.com/jnm2/9efbdd8ca0370630e6e812b1c47cd792)

I think this is the bootstrapper I'll use from now on. It's 49% of the length (complexity?) of the current bootstrapper and it respects `#cake version=0.24.0` and `#cake dir=mytools\Cake source="\\server\local packages"` in a systematic series of scenarios I've just tested. Of course if you decide to go with `#cake 0.24.0` I'll update it to work with that. The great thing is once I start using it, I'll be able to drop updates to it in any project and know for sure that I won't accidentally break a customization. I'll have to do some learning for the build.sh version.

Thanks for pinging me. I'm not sure why I'm having a hard time being convincing about the upsides of changing course and considering the bootstrappers a black-box arm of Cake, but I'll get out of your hair. 😊 
I am still very curious to know why, besides having made that choice in the past, you say the bootstrapper is supposed to be customized?

**Comment by devlead** [comment]
Hmm unsure, think something like below probably would be less complex and less error prone

### Bash
```bash
#!/usr/bin/env bash
CAKE_VERSION=0.25.0
SCRIPT_DIR=$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )
TOOLS_DIR=$SCRIPT_DIR/tools
CAKE_EXE=$TOOLS_DIR/Cake.$CAKE_VERSION/Cake.exe

# Make sure the tools folder exist.
if [ ! -d "$TOOLS_DIR" ]; then
  mkdir "$TOOLS_DIR"
fi

###########################################################################
# INSTALL CAKE
###########################################################################

if [ ! -f "$CAKE_EXE" ]; then
    echo "Installing Cake $CAKE_VERSION..."
    curl -Lsfo Cake.zip "https://www.nuget.org/api/v2/package/Cake/$CAKE_VERSION" && unzip -q Cake.zip -d "$TOOLS_DIR/Cake.$CAKE_VERSION" && rm -f Cake.zip
    if [ $? -ne 0 ]; then
        echo "An error occured while installing Cake."
        exit 1
    fi
fi

# Make sure that Cake has been installed.
if [ ! -f "$CAKE_EXE" ]; then
    echo "Could not find Cake.exe at '$CAKE_EXE'."
    exit 1
fi

###########################################################################
# RUN BUILD SCRIPT
###########################################################################

# Start Cake
(exec mono "$CAKE_EXE" --bootstrap) && (exec mono "$CAKE_EXE" "$@")
```

### PowerShell
```PowerShell
$CakeVersion = "0.25.0"

# Make sure tools folder exists
$PSScriptRoot = Split-Path $MyInvocation.MyCommand.Path -Parent
$ToolPath = Join-Path $PSScriptRoot "tools"
if (!(Test-Path $ToolPath)) {
    Write-Verbose "Creating tools directory..."
    New-Item -Path $ToolPath -Type directory | out-null
}

###########################################################################
# INSTALL CAKE
###########################################################################

Add-Type -AssemblyName System.IO.Compression.FileSystem
Function Unzip
{
    param([string]$zipfile, [string]$outpath)

    [System.IO.Compression.ZipFile]::ExtractToDirectory($zipfile, $outpath)
}


# Make sure Cake has been installed.
$CakePath = Join-Path $ToolPath "Cake.$CakeVersion/Cake.exe"
if (!(Test-Path $CakePath)) {
    Write-Host "Installing Cake $CakeVersion..."
     (New-Object System.Net.WebClient).DownloadFile("https://www.nuget.org/api/v2/package/Cake/$CakeVersion", "$ToolPath\Cake.zip")
     Unzip "$ToolPath\Cake.zip" "$ToolPath/Cake.$CakeVersion"
     Remove-Item "$ToolPath\Cake.zip"
}

###########################################################################
# RUN BUILD SCRIPT
###########################################################################

& "$CakePath" --bootstrap
if ($LASTEXITCODE -eq 0)
{
    & "$CakePath" $args
}
exit $LASTEXITCODE
```

**Comment by jnm2** [comment]
I'm still not going to go for something that duplicates the pinned version across two files. I've just run into too much recent pain, and it's just not as nice.

Unless the plan is for that version of Cake to be smart enough to bootstrap itself to a _canonical_ pinned version specified in the #cake directive in build.cake. Then the worst case scenario is that we'll always be downloading two copies of Cake every time you clone or clean your source control. That's not so bad.

**Comment by devlead** [comment]
> I'm still not going to go for something that duplicates the pinned version across two files

I do see your point and it's a very valid one, but from experience I'm a bit burned on supporting too complex boostrapppers. As bash & PowerShell are vastly different, people run different versions of them, it's been a hassle to keep them in sync between platforms, languages and new Cake features.

We've hit edge cases so many times - and it's frankly felt it being a bit out of scope of the project to cater for every possible scenario. That's why I feel it's better for the script to just download Cake, use it's boostrapping features for modules and just proxy args to it, that way we get feature parity out of the box with any platform, and makes it easier to support more platforms.

If we would to add more magic to the process, then it IMHO should be in the Cake bootstrapping command, which would inherently become xplat as it's in the runner itself and C#.

I don't think we should try downloading something each time as we do want to cater of offline scenarios and it's also a potential transient error source.

I'm willing to be convinced and overruled either way, but I also do feel the less magic/code in the bootstrappers the less can go wrong, the easier it becomes to have a consistent experience between platforms, which not only makes it easier to support existing and coming platforms, but also easier to document as it's one way regardless which shell you're using.

**Comment by jnm2** [comment]
Well put.
So long as the bootstrapper passes --bootstrap to Cake, or Cake bootstraps itself by default when it sees a pinned version that doesn't match its current version, I'm satisfied with the above for the sake of projects in general that use Cake.

I still may use my version that downloads the pinned version of Cake on the first shot, but at that point it's a very minor difference from the official bootstrapper.

I've seen version pinning proposed via a `#cake 0.24.0`-style directive. When I had thought about this before talking at Gitter, my mind went there first too. Then I realized that, since I want to do all my configuration in the single .cake file as I've stated before, I'm going to ask that every setting that you *currently* expose through cake.config be exposed also through a directive for people like me who need to see the script entry point setup all in one place. You *might* want to add more directives in the future. Would you consider going with `#cake version=0.24.0` instead, or something that leaves room for non-version settings?

Here's the immediate concrete reason I ask: my bootstrapper can't just invent a directive like `#cakedir mytools` or the build script will fail to compile. `#cake dir=mytools` is nice IMO because Cake can ignore it so the build script runs but my bootstrapper can still understand it. If you don't allow some similar syntax, I'm going to have to use `//#cakedir mytools` as the only possible way for the build script to still compile. What do you think?

**Comment by devlead** [comment]
@jnm2 you would need to do  a module, if you would like to add your own preprocessor directives

Adding a config directive is an interesting idea at first glance, but it does pose some real challenges though, as the config is needed before bootstrapping and script is executed.

**Comment by jnm2** [comment]
Cake is already reading the .config file, so why not have it read both the .config file and the config directives from the script in the same initial pass which it is already doing? Along with environment variables etc.

**Comment by devlead** [comment]
@jnm2 if you want to "abuse" something 😆  You could always use something undefined, i.e. this will compile fine
```cake
#if (NOT_RUNNING_CAKE)
{
    "Cake":  {
                 "Version":  "0.25.0"
             }
}
#endif
Information("Hello!");
```

**Comment by jnm2** [comment]
😆 Thought of that, but one-liners are really where my hope lies!

**Comment by devlead** [comment]
> Cake is already reading the .config file, so why not have it read both the .config file and the config directives from the script in the same initial pass which it is already doing? Along with environment variables etc.

Anything's possible, if it's an good idea could be a different thing:) would mean we would need to parse the cake script twice(4 times with bootrapping), not ruling anything out but I can see it becoming very confusing very fast.

**Comment by devlead** [comment]
> The only thing I'm really asking for is the ability for Cake to ignore key-value pairs with unknown key names.

There's a clear risk that by allowing anything, will result in that we can't add anything without risking breaking existing scripts, as that would allow infinite user directives.

> To me the important part is the result for the end user.

To me it's also important to be able to maintain and reason about the code in a sustainable way. Solutions that seem simple, can easily become a burden to maintain.

**Comment by jnm2** [comment]
> There's a clear risk that by allowing anything, will result in that we can't add anything without risking breaking existing scripts, as that would allow infinite user directives.

I don't think that risk is any greater than unknown key names in the cake.config file. In fact, it's a necessity to allow unknown key names. Cake 0.25.0 can't tell the difference between Cake 0.26.0 key names and something a user made up. I think that has to be okay! 😄

> There's a clear risk that by allowing anything, will result in that we can't add anything without risking breaking existing scripts, as that would allow infinite user directives.

That's your prerogative as a maintainer. I'm making the case as an end user knowing that the decision is yours. :+1:

**Comment by devlead** [comment]
Just a note this is a PoC PowerShell script that seems to work 2.0 to PS6 Core, tested on Windows & MacOS
```PowerShell
$CakeVersion = "0.25.0"
$IsRunningOnUnix = [System.Environment]::OSVersion.Platform -eq [System.PlatformID]::Unix
# Make sure tools folder exists
$PSScriptRoot = Split-Path $MyInvocation.MyCommand.Path -Parent
$ToolPath = Join-Path $PSScriptRoot "tools"
if (!(Test-Path $ToolPath)) {
    Write-Verbose "Creating tools directory..."
    New-Item -Path $ToolPath -Type directory | out-null
}

###########################################################################
# INSTALL CAKE
###########################################################################
if (-not (Get-Command Expand-Archive -ErrorAction SilentlyContinue))
{
    if ($PSVersionTable.PSVersion.Major -le 3)
    {
        & {
            function global:Expand-Archive
            {
                param([string]$Path, [string]$DestinationPath)
                $shell = New-Object -com shell.application
                $zip = $shell.NameSpace($Path)
                foreach($item in $zip.items())
                {
                    $shell.Namespace($DestinationPath).copyhere($item)
                }  
            }
        }      
    }
    else
    {
        & {
            Add-Type -AssemblyName System.IO.Compression.FileSystem
            function global:Expand-Archive
            {
                param([string]$Path, [string]$DestinationPath)
                [System.IO.Compression.ZipFile]::ExtractToDirectory($Path, $DestinationPath)
            }
        }
  }
}

# Make sure Cake has been installed.
$CakePath = Join-Path $ToolPath "Cake.$CakeVersion" 
$CakeExePath = Join-Path $CakePath "Cake.exe"
$CakeZipPath = Join-Path $ToolPath "Cake.zip"
$CakeUri = "https://www.nuget.org/api/v2/package/Cake/$CakeVersion"
if (!(Test-Path $CakeExePath)) {
    Write-Host "Installing Cake $CakeVersion..."
    (New-Object System.Net.WebClient).DownloadFile($CakeUri, $CakeZipPath)
    Expand-Archive $CakeZipPath $CakePath
    Remove-Item $CakeZipPath
}

###########################################################################
# RUN BUILD SCRIPT
###########################################################################
if ($IsRunningOnUnix)
{
    & mono "$CakeExePath" --bootstrap
    if ($LASTEXITCODE -eq 0)
    {
        & mono "$CakeExePath" $args
    }
}
else
{
    & "$CakeExePath" --bootstrap
    if ($LASTEXITCODE -eq 0)
    {
        & "$CakeExePath" $args
    }
}
exit $LASTEXITCODE
```

**Comment by jnm2** [comment]
[Since I linked it before– found and fixed some relative-path bugs in my [gist](https://gist.github.com/jnm2/b697d803866635ed5375dc5fb0ad4f4c). I'll have to set up some CI.]

**Comment by jnm2** [comment]
[Since I linked it before– found and fixed some relative-path bugs in my [gist](https://gist.github.com/jnm2/f40b17baa95cc72e406c8eaf1d0bd456). I'll have to set up some CI. Also added the ability to detect `// #cake version=0.24.0` for the time being.]

**Comment by jnm2** [comment]
[Since I linked it before– found and fixed some relative-path bugs in my [gist](https://gist.github.com/jnm2/6271d85234163ed257b9125bd7c738aa). I'll have to set up some CI. Also added the ability to detect `// #cake version=0.25.0` for the time being. [Example of .cake file](https://github.com/jnm2/Elevate/blob/master/build.cake#L1) using this bootstrapper.]

**Comment by damianh** [comment]
Sorry to butt in ... am here with @pblasucci discussing this very topic. We'd like to share that we (think) we would like a simple native cli tool (prob written in `go` lol) that does the bootstrapping work.

**Comment by gep13** [comment]
@damianh hmmm... interesting...

**Comment by devlead** [comment]
@damianh interesting, please do tell more, would need a native binary for each target platform.

So would need either a minimal build.sh/ps1 passing something like `uname -a`  or `[System.Environment]::OSVersion`to an API to download right binary, or?

**Comment by damianh** [comment]
Yes a native bin for each platform would be needed. I would commit to the repo instead of download (although that is an option too with what you have suggested.)

**Comment by devlead** [comment]
@damianh One thought, isn't static linked go binaries fairly large? Would make many repos considerably fatter.At least 6mb + any revision changes by looking at golang/go#6853 
Personally I generally feel a bit reserved about adding binaries to all my repos.

**Comment by patriksvensson** [comment]
I'm going to take a stab at this writing a simple proof of concept using Rust if everyone's OK with it.

**Comment by devlead** [comment]
@patriksvensson go nuts, great if we can have a few different PoC and ideas.

**Comment by damianh** [comment]
@patriksvensson That would be very interesting to see.

@devlead I don't hold my suggestion strongly. I just know that I have many variations of a bootsrappers scattered across repos at the moment and it's a bit irksome. Open to other ideas.

**Comment by devlead** [comment]
> I don't hold my suggestion strongly. 

@damianh me neither :)

People clearly have different needs to, initial goal was to get people going quickly with as few dependencies as possible.  The "smarter" Cake gets the dumber the bootstrapper can be.

But great that we're having this discussion so we can bounce ideas and get to a good middle ground.

Personally I prefer if the bootstrappers act and feel as similar as possible across OS:se and runtimes. IMHO, bootstrappers doing too much has caused confusion and inconsistencies in the past, i.e. argument parsing errors and bootstrappers making stuff don't work - which work fine if you invoke Cake directly.

So welcome all ideas and feedback we can get to stream line the experience as much as possible, while not impairing the tool or creating too much OS / Runtime specific thingies.

**Comment by daveaglick** [comment]
With the announcement of .NET Core 2.1 it looks like global tools are going to ship (https://blogs.msdn.microsoft.com/dotnet/2018/02/02/net-core-2-1-roadmap/). That means it may be possible to have an experience like:
```
dotnet tool install -g cake
cake -target DoYourThing
```
In my mind, it also raises the question of whether a bootstrapper is needed at all for .NET Core. Just make Cake itself a global tool and let the dotnet CLI and NuGet bootstrap it. On the one hand this approach certainly looks simple at first glance, though there may be complications with things like versioning. On the other hand it may mean maintaining two different approaches depending on .NET Core vs. Framework/Mono.

**Comment by patriksvensson** [comment]
@damianh @jnm2 @mholo65 @devlead @gep13 

I've created a proof of concept I (for now) call `cakeup` (like rustup): https://github.com/patriksvensson/cakeup

Seems to work nicely and the statically linked binary is around 700kb which isn't super big, but we can probably trim this even further.

**Comment by patriksvensson** [comment]
Created an example of how an application can be bootstrapped with [cakeup].
https://github.com/patriksvensson/cakeup-example

PowerShell: https://github.com/patriksvensson/cakeup-example/blob/master/build.ps1
Bash: https://github.com/patriksvensson/cakeup-example/blob/master/build.sh

**Comment by damianh** [comment]
oh nice, will give this a spin!

**Comment by nalla** [comment]
@patriksvensson the cakeup link is dead. Did it move?

**Comment by devlead** [comment]
It's still WIP,  repo is here
https://github.com/cake-build/cakeup

**Comment by pascalberger** [comment]
Closing issue since recommended approach nowadays is to use the Cake .NET tool, which has a much more simple bootstrapping process. See https://github.com/cake-build/resources/issues/95 for providing an example bootstrapping script for it in this repository.
